Mathlib/Topology/UniformSpace/Ultra.lean

```diff
@@ -0,0 +1,129 @@
+/-
+Copyright (c) 2017 Yakov Pechersky. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yakov Pechersky
+-/
+import Mathlib.Topology.UniformSpace.Defs
+import Mathlib.Topology.Bases
+
+/-!
+# Ultrametric (nonarchimedean) uniform spaces
+
+Ultrametric (nonarchimedean) uniform spaces are ones that generalize ultrametric spaces by
+having a uniformity based on equivalence relations.
+
+## Main definitions
+
+In this file we define `IsUltraUniformity`, a Prop mixin typeclass.
+
+## Main results
+
+* `TopologicalSpace.isTopologicalBasis_clopens`: a uniform space with a nonarchimedean uniformity
+  has a topological basis of clopen sets in the topology, meaning that it is topologically
+  zero-dimensional.
+
+## Implementation notes
+
+As in the `Topology/UniformSpace/Defs.lean` file, we do not reuse `Data/Rel.lean`
+but rather extend the relation properties as needed.
+
+## TODOs
+
+* Prove that `IsUltraUniformity` iff `IsUltrametricDist` on a `PseudoMetricSpace`
+* Generalize `IsUltrametricDist` to `IsUltrametricUniformity`
+* Provide `IsUltraUniformity` for the uniformity in a `Valued` ring
+* Generalize results about open/closed balls and spheres in `IsUltraUniformity` to
+  combine applications for `MetricSpace.ball` and valued "balls"
+* Use `IsUltraUniformity` to work with profinite/totally separated spaces
+* Define the nonarchimedean uniformity of a space that is a product of `IsUltraUniformity`s
+* Show that the `UniformSpace.Completion` of an `IsUltraUniformity` is `IsUltraUniformity`
+
+## References
+
+* [D. Windisch, *Equivalent characterizations of non-Archimedean uniform spaces*][windisch2021]
+* [A. C. M. van Rooij, *Non-Archimedean uniformities*][vanrooij1970]
+
+-/
+
+open Set Filter Topology
+open scoped Uniformity
+
+variable {X : Type*}
+
+/-- The relation is transitive. -/
+def IsTransitiveRel (V : Set (X × X)) : Prop :=
+  ∀ ⦃x y z⦄, (x, y) ∈ V → (y, z) ∈ V → (x, z) ∈ V
+
+lemma IsTransitiveRel.comp_eq_of_idRel_subset {s : Set (X × X)}
+    (h : IsTransitiveRel s) (h' : idRel ⊆ s) :
+    s ○ s = s := by
+  refine le_antisymm ?_ (subset_comp_self h')
+  intro ⟨x, y⟩
+  simp only [mem_compRel, forall_exists_index, and_imp]
+  intro z
+  exact @h x z y
+
+variable [UniformSpace X]
+
+variable (X) in
+/-- A uniform space is ultrametric if the uniformity `𝓤 X` has a basis of equivalence relations. -/
+class IsUltraUniformity : Prop where
+  has_basis : (𝓤 X).HasBasis
+    (fun s : Set (X × X) => s ∈ 𝓤 X ∧ IsSymmetricRel s ∧ IsTransitiveRel s) id
+
+variable [IsUltraUniformity X]
+
+namespace UniformSpace
+
+lemma nhds_basis_clopens (x : X) :
+    (𝓝 x).HasBasis (fun s : Set X => x ∈ s ∧ IsClopen s) id := by
```

Comment:
Could you break this down into:
- balls for transitive relations are open
- balls for equivalence relations are disjoint (or equal)
- balls for equivalence relations are clopen
- and then this result follows almost immediately using [nhds_basis_uniformity'](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/UniformSpace/Defs.html#nhds_basis_uniformity') (more precisely this gives you *a* basis of clopen sets, and then [Filter.HasBasis.to_hasBasis'](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Order/Filter/Bases/Basic.html#Filter.HasBasis.to_hasBasis') finishes the job)

Reply:
I did most of this in [323d893](https://github.com/leanprover-community/mathlib4/pull/22580/commits/323d89349cc98d22945baaddf50acd78bb6eb606)

I didn't do the `Disjoint` lemma because really that should go through `[disjoint_or_subset_of_isClopen](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Connected/Clopen.html#disjoint_or_subset_of_isClopen)` iiuc.

Reply:
No I didn't mean [disjoint_or_subset_of_isClopen](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Connected/Clopen.html#disjoint_or_subset_of_isClopen), because this has a connectedness hypothesis. But after playing a bit with the proof (see above suggestion) I realize that the disjointness lemma would essentially be unnecessary boilerplate.